aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -0,0 +1,258 @@
+#include <ATen/native/Indexing.h>
+#include <ATen/native/IndexingUtils.h>
+
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/AccumulateType.h>
+
+#include <THC/THCDeviceUtils.cuh>
+#include <THC/THCGeneral.h>
+#include <THC/THCTensorSort.cuh>
+#include <ATen/cuda/CUDAContext.h>
+#include <THC/THCThrustAllocator.cuh>
+#include <thrust/execution_policy.h>
+#include <thrust/sort.h>
+
+
+namespace {
+
+#ifdef __HIP_PLATFORM_HCC__
+static const int WARP_SIZE = 64;
+#else
+static const int WARP_SIZE = 32;
+#endif
+
+
+
+
+template <typename scalar_t, int SZ>
+__global__ void indexing_backward_kernel(
+  int64_t* input, int64_t* indices, scalar_t* grad_output, scalar_t* grad_weight,
```

Comment:
I find the use of variable name `input` here kind of confusing, because, based on the call site, this isn't actually an input in that sense; it's `sorted_indices`, no?